### ISSUE TITLE
Unable to extend virtual dependencies

### ISSUE THREAD

**Issue by adamjstewart** [LLNL/spack#17475]
Two core tenets of Spack are the idea of virtual packages (where multiple implementations of a library provide the same API) and extensions (where a package can be symlinked to a dependency prefix). Unfortunately, these two tenets are currently incompatible, and virtual dependencies cannot be extended.

### Rationale

There are many extendable packages in Spack that either have virtual dependencies or would benefit from having them. For example:

Java has [several implementations](https://en.wikipedia.org/wiki/List_of_Java_virtual_machines):

- Oracle JDK
- OpenJDK
- IBM Java
- IcedTea

Java is already a virtual package in Spack.

Python has [several implementations](https://www.python.org/download/alternatives/):

- CPython
- Anaconda
- Miniconda
- PyPy
- Jython
- IronPython
- Intel Python

Many users have requested that Python become a virtual dependency (#7966).

### Description

I first noted this behavior in #8613, but for a virtual dependency like `java`, if you write:
```python
extends('java')
```
in a package, concretization will crash with the following error message:
```
==> Error: Package 'java' not found.
```
For now, you have to write `extends('jdk')` if you want it to work. Of course, this prevents users from using `openjdk` or any other Java provider. This problem will become even more apparent once we make Python a virtual dependency.

### Additional information

There are several other issues that are related to this issue that we should also solve:

- Unable to extend multiple packages (#987)
- Virtual dependencies do not support variants (#1712, #6300)
- Extensions do not support variants (#10195)
- Multiple providers found for virtual dependency (#4107, #7339, #7901, #8059, #8133, #9753, #11987, #13821, #20308)

**Comment by tgamblin** [comment]
Thanks for making this!

**Comment by becker33** [comment]
@adamjstewart @tgamblin

I think we can now work around this constraint with the newer features in the concretizer. Specifically:

```
class CPython(...):
    versions = [
        ("3.9.2", {"sha256": "...", ...}
    ]
    for version, args in versions:
        version(version, **args)
        provides(f"python@{version}", when=f"@{version}")
```
```
class PythonExtension(...):
    ...
    depends_on("python")
    for name in spack.repo.path.providers_for("python"):
        extends(name, when=f"^{name}")
    ... 
```
What do y'all think about moving forward with this? This was prompted by the discussion in #38823

**Comment by adamjstewart** [comment]
> ```python
> depends_on("python")
> for name in spack.repo.path.providers_for("python"):
>     extends(name, when=f"^{name}")
> ```

PythonExtension doesn't currently depend on Python since it's optional for many packages. Similarly for Java, we don't have a JavaExtension base class. So we would need to copy that for-loop for all packages that extend Java and many packages that extend Python. Is there any way we could move that for-loop inside of the `extends` statement?

> ```python
> for version, args in versions:
>     version(version, **args)
>     provides(f"python@{version}", when=f"@{version}")
> ```

We've been talking about adding a parameter to `depends_on` to support the common use case:
```python
for ver in versions:
    depends_on(f"foo@{ver}", when=f"@{ver}")
```
See #14002 and #37418. It would also be nice if we had a similar parameter for `provides`. This could be used for CPython, Pillow, JDK, and a lot of other packages.

**Comment by becker33** [comment]
> PythonExtension doesn't currently depend on Python since it's optional for many packages.

Yes, I was perhaps too abbreviated in my shortened example. The actual code would of course have to look something like:

```
    with spack.multimethod.when("build_system=python_pip"):                                                                                                                                    
        depends_on("python")
        for name in spack.repo.path.providers_for("python"):
            extends(name, when=f"^{name}")
```
which mirrors the current code in `PythonExtension`

I don't off the top of my head see a reason we couldn't modify the `extends` directive to do the same under the hood if the extendee is a virtual. The current directive is

```
@directive(("extendees", "dependencies"))
def extends(spec, type=("build", "run"), **kwargs):
    """Same as depends_on, but also adds this package to the extendee list.

    keyword arguments can be passed to extends() so that extension
    packages can pass parameters to the extendee's extension
    mechanism.

    """

    def _execute_extends(pkg):
        when = kwargs.get("when")
        when_spec = make_when_spec(when)
        if not when_spec:
            return

        _depends_on(pkg, spec, when=when, type=type)
        spec_obj = spack.spec.Spec(spec)
        pkg.extendees[spec_obj.name] = (spec_obj, kwargs)

    return _execute_extends
 ```

We could do something like:

```
@directive(("extendees", "dependencies"))
def extends(spec, type=("build", "run"), **kwargs):
    """..."""

    def _execute_extends(pkg):
        when = kwargs.get("when")
        when_spec = make_when_spec(when)
        if not when_spec:
            return

        _depends_on(pkg, spec, when=when, type=type)
        spec_obj = spack.spec.Spec(spec)
        if spec_obj.virtual:
            for provider in spack.repo.path.providers_for(str(spec_obj)):
                when = when_spec.constrained(f"^{provider}")
                local_kwargs = kwargs.copy()
                local_kwargs["when"] = when
                provider_obj = spack.spec.Spec(provider)
                pkg.extendees[provider_obj.name] = (provider_obj, local_kwargs)
        else:
            pkg.extendees[spec_obj.name] = (spec_obj, kwargs)

    return _execute_extends
```

**Comment by becker33** [comment]
Ok playing with this some I'm reminded of the /*other*/ reason we can't have a python virtual yet -- we need variants on the virtual interface for cpython variants that affect functionality.

I'm less familiar with pypy, so I don't know how they are implemented there. I've been playing with skybison as an easy-to-build alternative python for now.

**Comment by becker33** [comment]
The good news is the extension portion of this appears to be working pretty much as-written above.

**Comment by adamjstewart** [comment]
I can add a PyPy package. For now, Java is probably easier to experiment with.

**Comment by adamjstewart** [comment]
Bleh, PyPy requires bootstrapping with itself. Trying to decide if I want to write a quick binary-only package or just ignore this and we can experiment with using conda as a different Python provider.

**Comment by becker33** [comment]
I thought pypy can also be built with cpython?

**Comment by adamjstewart** [comment]
Oh maybe, I didn't think about that. But this opens up another can of worms. Virtuals can't depend on other virtuals, correct? A single virtual can only occur once in the DAG. Like if both CPython and PyPy are in the DAG, the concretizer would not be happy.

**Comment by becker33** [comment]
@alalazo is working on relaxing that for build dependencies

**Comment by alalazo** [comment]
@adamjstewart The PR is #38447 currently waiting for review. It will be my top priority to get it merged, once it's reviewed.
